batch-processing/README.md

```diff
@@ -0,0 +1,379 @@
+
+The tool should (in order of priority):
+
+1. Write thorough logs (`log.Println` and `log.Printf`) to describe what it is doing, including errors
+1. Validate the input CSV to ensure it only has **one** column, `url`
+1. Gracefully handle failures & continue to process the input CSV even if one row fails
+1. Support a configurable AWS region and S3 bucket via environment variables `AWS_REGION` and `S3_BUCKET`
+
+### Project extension
+
+To take this project further, add these requirements:
+
+1. The tool should not upload the same image twice
+1. The tool should not process the same image twice
+1. The tool should process and upload in parallel, using [goroutines](https://go.dev/tour/concurrency/1)
+
+## How-to
+
+Most of getting this project build is up to you. However, here are some pointers for things you are going to need.
+
+### Reading a CSV
+
+The built-in [`encoding/csv`](https://pkg.go.dev/encoding/csv) package is the one to use to read and write the CSV files.
+
+### Image processing
+
+Most of the ImageMagick code, which grayscales the image, is written for you. This shouldn't change too much.
+
+### Cloud storage (S3)
+
+The specification asks you to upload the images to [Amazon AWS S3 cloud storage](https://aws.amazon.com/s3/). S3 is a cloud service designed to store huge amounts of data and files in a secure, scalable and cost-effective way.
+
+The basic organisation of S3 is simple. At the top level of S3 is the **bucket**. Data ("objects") are stored within these buckets in folders, like a normal file system. Read the [high-level guide to S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html).
+
+You will need to create a bucket — call it anything you like. Pay attention to which [AWS region](https://cloudacademy.com/blog/aws-regions-and-availability-zones-the-simplest-explanation-you-will-ever-find-around/#:~:text=What%20are%20AWS%20Regions%3F,host%20their%20cloud%20infrastructure%20there.) you create it in – you will need this later.
+
+You can make it publically accessible to the internet, so that you or anyone can view the images that are uploaded, with a policy like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "PublicAccess",
+      "Effect": "Allow",
+      "Principal": "*",
+      "Action": ["s3:GetObject", "s3:GetObjectVersion"],
+      "Resource": "arn:aws:s3:::[NAME OF BUCKET]/*"
+    }
+  ]
+}
+```
+
+> ⚠️ Public means public. **DO NOT** upload anything sensitive to the bucket.
+
+Use the [AWS SDK for Go](https://github.com/aws/aws-sdk-go) to interact with Amazon S3.
+
+You will need to set up credentials so that your docker image can write to S3. The best way to do that is to:
+
+- Create a role + policy to allow uploads to the bucket
+- Allow your AWS user account to use the role
+- Make your AWS credentials available to the Docker image
```

Comment:
```suggestion
- Make your AWS credentials available to the Docker container
```

I'd generally lean towards "mount secrets at runtime" rather than "bake secrets into an image" :)

Reply:
Ooft, yes. This was my intent, wrong word. Thanks.